docs/enhancements/automatic-agent-classification-labels.md

```diff
@@ -0,0 +1,101 @@
+
+## Motivation
+
+Having users define labels manually is time-consuming and error-prone.  Defining conditions is more reliable and scalable for users.
+
+### Goals
+
+- Provide a flexible and easy-to-use API for defining desired labels and conditions.
+- Extend the Infrastructure Operator to maintain the desired labels based on the specified conditions with minimal impact on scale.
+
+### Non-Goals
+
+- Ship pre-defined labels and conditions.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Cluster Creator, I want the system to automatically label Agents based on specific conditions so that I can more easily select Agents for cluster membership.	
+
+## Design Details [optional]
+
+We will create a new AgentClassification namespace-scoped CRD that defines the desired label key-value and a list of expressions to be evaluated according to the Agent’s inventory.  Each expression is generic such that it can refer to any properties in the inventory.
+
+### AgentClassification CRD Definition
+
+```
+// AgentClassificationSpec defines the desired state of AgentClassification
+type AgentClassificationSpec struct {
+	// LabelKey specifies the label key to apply to matched Agents
+	//
+	// +immutable
+	LabelKey string `json:"labelKey"`
+
+	// LabelValue specifies the label value to apply to matched Agents
+	//
+	// +immutable
+	LabelValue string `json:"labelValue"`
+
+	// ExpressionDefinitions define the conditions upon which the specified label is applied
+	ExpressionDefinitions []ExpressionDefinition `json:"expressionDefinitions"`
+}
+
+// ExpressionDefinition defines an expression to be evaluated
+type ExpressionDefinition struct {
+	// Expression is the expression that will be evaluated, in govaluate format
+	Expression string `json:"expression"`
+
+	// Parameters maps parameters from the above expression to paths in the Agent inventory
+	Parameters map[string]string `json:"parameters"`
+}
+```
+
+We will base the expression definition on the [govaluate](https://github.com/Knetic/govaluate) library, which supports a language for expressions with parameters.  Along with each expression (e.g., “cpu > 8”), the user will define a list of variables and map them to inventory items (e.g., “cpu”: “Cpu.Count”).  The controller will need to parse the inventory path and translate it into the corresponding value.  In addition, the inventory contains arrays for disks and network interfaces - the user will pass "Disks(N).SizeBytes" for at least N disks having the specified capacity, or "Disks().SizeBytes" for all disks.  The controller will need to parse this and create additional govaluate expressions for each disk, and then check for at least N returning true.
+
+### Operator changes
+
+When an Agent's inventory is updated, the Agent controller will now need to list all AgentClassification CRs in the namespace and update the Agent's labels accordingly.
```

Comment:
does it have to be the same controller? 
If i remember correct when updating inventory agent controller doesn't look at inventory changes, maybe it need to be a different controller that will watch the agents?

Reply:
Why do we need another controller to watch agents?  We already have a method that updates the inventory in the Status, so why not update the labels there?

Reply:
Actually it depend on the scale as well, if we have the agent to do all the work then it will mean that the reconcile loop will be mo complicated and will probably run longer, if it's a different controller then we can split the work, after all it is just adding labels. 
I'm not sure if it's really the right way, but it looks like it's doing very specific work.

Reply:
Good question.  I'm not sure how much parallelism there is between controllers, so even if we move it to another controller they might execute serially.  I will check that.